core/server/worker/src/main/java/alluxio/worker/netty/BlockDataServerHandler.java

```diff
@@ -115,6 +116,52 @@ void handleBlockReadRequest(final ChannelHandlerContext ctx, final RPCBlockReadR
   }
 
   /**
+   * Handles a {@link RPCUnderFileSystemBlockReadRequest} by reading the data through a
+   * {@link BlockReader} provided by the block worker. This method assumes the data is available
+   * in the UFS returns an error status if the data is not available.
+   *
+   * @param ctx The context of this request which handles the result of this operation
+   * @param req The initiating {@link RPCBlockReadRequest}
+   * @throws IOException if an I/O error occurs when reading the data requested
+   */
+  public void handleUnderFileSystemBlockReadRequest(final ChannelHandlerContext ctx,
+      final RPCUnderFileSystemBlockReadRequest req) throws IOException {
+    final long blockId = req.getBlockId();
+    final long offset = req.getOffset();
+    final long len = req.getLength();
+    final long sessionId = req.getSessionId();
+    final boolean noCache = req.getNoCache();
+
+    try {
+      DataBuffer buffer = null;
+      req.validate();
+      BlockReader reader = mWorker.readUfsBlock(sessionId, blockId, offset, noCache);
+      ByteBuffer data = reader.read(offset, len);
```

Comment:
yes, this is guaranteed in our existing client code (if UnknownLengthFileInStream is not used) though we can write a client that sends a request with length greater than the block size.